lib/frontend.ml

```diff
@@ -163,7 +164,12 @@ module Make(C: S.CONFIGURATION with type 'a io = 'a Lwt.t) = struct
       let pages = Io_page.pages num in
       List.iter
         (fun (gref, page) ->
-           let id = gref mod (1 lsl 16) in
+           let rec next () =
+             let id = nf.rx_id in
+             nf.rx_id <- (succ nf.rx_id) mod (1 lsl 16) ;
+             if Hashtbl.mem nf.rx_map id then next () else id
```

Comment:
Around 300 elements in my experimental setup.  What I wanted here is a minimal invasive patch. Better safe than sorry.

I do think that in the long run using a shared pool of pages, as suggested by talex5, is the way to go (which also avoids issuing grants for each page), but will need copying into another buffer before passing up the stack (afaict).  I don't feel confident enough with the mirage-net-xen codebase to develop this (rather big) change, and will leave this to the original authors for future versions.